[Feature] Pools table bookmarks

yonikid15 opened the pull request:
🤖 Resolves #10269

## 👋 Introduction

**This is a draft until the ordering works correctly.**

- Adds bookmarking feature to pools table.
- Adds a pivot table btw pool and users (bookmarks).
- The bookmarks are unique to each user.
- The bookmarks should stick to top of table (still working on this).

## 🕵️ Details

Add any additional details that could assist with reviewing or testing this PR.

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Build app
2. Login as admin
3. Go to view processes on admin dashboard
4. Ensure bookmarking feature works along with other table tools.

vd1992 commented:
Where is sorting failing? It seems to work for me. 
Doesn't appear to do a secondary sort though, sorting a second way after bookmark, so sort by bookmark and then say stream 

![Uploading image.png…]()

vd1992 commented on api/graphql/schema.graphql:
```diff
@@ -1953,6 +1963,13 @@ type Mutation {
     @update
     @guard
     @canFind(ability: "updatePublished", find: "id")
+  togglePoolUserBookmark(
+    userId: UUID! @rename(attribute: "user_id")
+    poolId: UUID! @rename(attribute: "pool_id")
+  ): Pool
+    @guard
+    @canFind(ability: "view", find: "pool_id", model: "Pool", injectArgs: true)
```
This allows anyone to update anyone else's bookmark which we probably don't want. I think it is reasonable for someone to only be able to update their own bookmarks.

I think that can be easily done like so below using `inject`, where the acting/auth context object feeds into the `args.user_id` value

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/2283baa7-6d26-438e-9172-c0b00b652155)

vd1992 reviewed (commented):
It seems to work well, I can't comment on if the sorting is ideal since I didn't get what was wrong. Did find something important below

vd1992 commented on api/graphql/schema.graphql:
```diff
@@ -1962,6 +1972,11 @@ type Mutation {
     @guard
     @canFind(ability: "updatePublished", find: "id")
+  togglePoolUserBookmark(poolId: UUID! @rename(attribute: "pool_id")): Pool
+    @guard
+    @inject(context: "user.id", name: "user_id")
```
You are injecting `user_id` here

yonikid15 replied:
See [79cc264](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/79cc264065ed10aee21f6ca7574f7031f338df8e)

vd1992 commented on api/app/GraphQL/Mutations/TogglePoolUserBookmark.php:
```diff
@@ -0,0 +1,25 @@
+     */
+    public function __invoke($_, array $args)
+    {
+        /** @var \App\Models\User */
+        $user = Auth::user();
+        $pool = Pool::find($args['pool_id']);
+        $user->poolBookmarks()->toggle($pool->id);
```
But do not use `user_id` anywhere

I don't know what the safe/best way would be though. The highlighted above or finding from the injected `user_id`
Any wisdom to share on this @tristan-orourke  ?

yonikid15 replied:
See [79cc264](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/79cc264065ed10aee21f6ca7574f7031f338df8e)

yonikid15 replied:
I talked with Tristan and he said this option looks good :+1:

vd1992 commented on api/graphql/schema.graphql:
```diff
@@ -815,10 +823,12 @@ type Query {
     orderBy: _
       @orderBy(
         relations: [
+          # { relation: "poolBookmarks", columns: ["created_at"] }
```
Commented out

vd1992 commented on apps/web/src/pages/Pools/IndexPoolPage/components/helpers.tsx:
```diff
@@ -190,14 +195,27 @@ export function getOrderByClause(
     // ["status", "status"],
   ]);
 
+  // const poolBookmarksOrderByClause: QueryPoolsPaginatedOrderByRelationOrderByClause =
+  //   {
+  //     column: undefined,
+  //     order: SortOrder.Asc,
+  //     poolBookmarks: {
+  //       aggregate: OrderByRelationWithColumnAggregateFunction.Max,
+  //       column: "CREATED_AT" as QueryPoolsPaginatedOrderByPoolBookmarksColumn,
+  //     },
+  //   };
```
Commented out

yonikid15 replied:
Fixed by [79cc264](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/79cc264065ed10aee21f6ca7574f7031f338df8e)

vd1992 reviewed (commented):
Got some commented out bits below, did you mean to keep them around?
I'd also like to let someone else handle final approval to judge sorting

yonikid15 commented:
Sorry Vachan I might have accidentally clicked the re-review.  However, I just addressed the comments. For the other comment, I decided to get the user form Auth, however we can see what Tristans comment is on that.

petertgiles commented on api/graphql/schema.graphql:
```diff
@@ -815,6 +823,7 @@ type Query {
     includeIds: [UUID!] @in(key: "id")
     excludeIds: [UUID!] @notIn(key: "id")
     where: PoolFilterInput
+    orderByPoolBookmarks: PoolBookmarksOrderByInput @scope
```
Why is a second query fired when the orderby is already satisfied in the first query?
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/a3d1d7b5-797b-4e25-ba0a-39333161cccb)

yonikid15 replied:
I wasn't seeing this, but let me double check.

petertgiles replied:
I'm still seeing it but if it's not obvious to you I think we can let it go for later optimization.

petertgiles reviewed (changes requested):
Very cool!  :sunglasses: This works very well already.  I left a few small ideas to level it up.  Let me know if you want to chat about any of them.

yonikid15 commented on api/app/Models/User.php:
```diff
@@ -201,6 +202,11 @@ public function pools(): HasMany
         return $this->hasMany(Pool::class);
     }
 
+    public function poolBookmarks(): BelongsToMany
```
Add seeder in [16e44b9](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/16e44b9a1d57925a0505ccdac1f6431891fff0ce)

yonikid15 commented on api/app/GraphQL/Mutations/TogglePoolUserBookmark.php:
```diff
@@ -0,0 +1,24 @@
+use App\Models\Pool;
+use Illuminate\Support\Facades\Auth;
+
+final class TogglePoolUserBookmark
```
Add test in [4b0b06b](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/4b0b06b760ba4c8e024369f7d7ed701488ec6cf9)

yonikid15 commented on api/app/Models/Pool.php:
```diff
@@ -468,6 +473,28 @@ public function scopeOrderByTeamDisplayName(Builder $query, ?array $args): Build
+        // order the pools so that the bookmarks connected to current user sticks to the top
+        if ($order && $user) {
+            $query->orderBy(
+                $user->select('pool_user_bookmarks.created_at')
```
I might need some more guidance on this. I removed take(1) and it works fine without it :+1: 
However, without ordering by the bookmarks `created_at` date, the order would vary causing the tests to fail.

petertgiles replied:
This is working OK for me, I think.  The order by created date in the orderBy parameter is keeping it steady.
![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/8978655/bb110e91-dddd-4b96-9876-8e18f151dc99)
We can look at this together if I'm missing something.

yonikid15 replied:
Awesome thanks this solution is working :+1:

petertgiles reviewed (approved):
Awesome!  :1st_place_medal:
